plugins/notion/src/api.ts

```diff
@@ -0,0 +1,464 @@
+    id: FieldId
+    name: string
+    originalName: string
+    type: ManagedCollectionField["type"] | null
+    allowedTypes: ManagedCollectionField["type"][]
+    notionProperty: NotionProperty | null
+}
+
+export type NotionProperty = GetDatabaseResponse["properties"][string]
+
+// Every page has content which can be fetched as blocks. We add it as a
+// property so it displays in the list where you can configure properties to be
+// synced with the CMS
+export const pageContentProperty: FieldInfo = {
+    id: "page-content",
+    type: "formattedText",
+    name: "Content",
+    originalName: "Content",
+    allowedTypes: ["formattedText"],
+    notionProperty: null,
+}
+
+// The valid field types that can be used as a slug, in order of preference
+const slugFieldTypes: NotionProperty["type"][] = ["title", "rich_text"]
+
+export const supportedCMSTypeByNotionPropertyType = {
+    checkbox: ["boolean"],
+    date: ["date"],
+    number: ["number"],
+    title: ["string"],
+    rich_text: ["formattedText", "string"],
+    created_time: ["date"],
+    last_edited_time: ["date"],
+    select: ["enum"],
+    status: ["enum"],
+    url: ["link"],
+    email: ["formattedText", "string"],
+    files: ["file", "image"],
+    relation: ["multiCollectionReference"],
+} satisfies Partial<Record<NotionProperty["type"], ReadonlyArray<ManagedCollectionField["type"]>>>
+
+// Naive implementation to be authenticated, a token could be expired.
+// For simplicity we just close the plugin and clear storage in that case.
+export function isAuthenticated() {
+    return localStorage.getItem(PLUGIN_KEYS.BEARER_TOKEN) !== null
+}
+
+let notion: Client | null = null
+if (isAuthenticated()) {
+    initNotionClient()
+}
+
+export function initNotionClient() {
+    const token = localStorage.getItem(PLUGIN_KEYS.BEARER_TOKEN)
+    if (!token) throw new Error("Notion API token is missing")
+
+    notion = new Client({
+        fetch: async (url, fetchInit) => {
+            const urlObj = new URL(url)
+
+            try {
+                const resp = await fetch(`${API_BASE_URL}/notion${urlObj.pathname}${urlObj.search}`, fetchInit)
+
+                // If status is unauthorized, clear the token
+                // And we close the plugin (for now)
+                // TODO: Improve this flow in the plugin.
+                if (resp.status === 401) {
+                    localStorage.removeItem(PLUGIN_KEYS.BEARER_TOKEN)
+                    await framer.closePlugin("Notion Authorization Failed. Re-open the plugin to re-authorize.", {
+                        variant: "error",
+                    })
+                    return resp
+                }
+
+                return resp
+            } catch (error) {
+                console.log("Notion API error", error)
+                throw error
+            }
+        },
+        auth: token,
+    })
+}
+
+export async function getNotionDatabases() {
+    if (!notion) {
+        initNotionClient()
+    }
```

Comment:
Turn `initNotionClient` into `getNotionClient` that just caches it in a `notionClientSingleton` variable or something like that, so that you don't need `notion!` and `assert` anymore.